examples/vpc-scenario-3/main.tf

```diff
@@ -0,0 +1,170 @@
+  release = "14.04"
+}
+
+resource "aws_key_pair" "main" {
+  key_name   = "${var.name}"
+  public_key = "${file(var.ssh_pubkey)}"
+}
+
+# Security group for the elastic load balancer
+module "elb-sg" {
+  source      = "../../modules/security-group-base"
+  description = "Allow public access to ELB in ${var.name}"
+  name        = "${var.name}-elb"
+  vpc_id      = "${module.vpc.vpc_id}"
+}
+
+# security group rule for elb open inbound http
+module "elb-http-rule" {
+  source            = "../../modules/single-port-sg"
+  port              = 80
+  cidr_blocks       = ["0.0.0.0/0"]
+  security_group_id = "${module.elb-sg.id}"
+  description       = "open HTTP on the ELB to public access"
+}
+
+# security group rule for elb open egress (outbound from nodes)
+module "elb-open-egress-rule" {
+  source            = "../../modules/open-egress-sg"
+  security_group_id = "${module.elb-sg.id}"
+}
+
+# Security group for the web instance, only accessible from ELB
+module "web-sg" {
+  source      = "../../modules/security-group-base"
+  description = "Allow HTTP and SSH to web instance in ${var.name}"
+  name        = "${var.name}-web"
+  vpc_id      = "${module.vpc.vpc_id}"
+}
+
+# allow HTTP from ELB to web instances
+module "web-http-elb-sg-rule" {
+  source            = "../../modules/single-port-sg"
+  port              = "3000"
+  description       = "Allow ELB HTTP to web app on port 3000"
+  cidr_blocks       = ["${module.vpc.public_cidr_blocks}"]
+  security_group_id = "${module.web-sg.id}"
+}
+
+# open egress for web instances (outbound from nodes)
+module "web-open-egress-sg-rule" {
+  source            = "../../modules/open-egress-sg"
+  security_group_id = "${module.web-sg.id}"
+}
+
+# Load Balancer
+resource "aws_elb" "web" {
+  name = "${var.name}-public-elb"
+
+  health_check {
+    healthy_threshold   = 2
+    interval            = 15
+    target              = "TCP:3000"
+    timeout             = "5"
+    unhealthy_threshold = 10
+  }
+
+  # public, or private to VPC?
+  internal = false
+
+  # route HTTPS to services app on port 3000
+  listener {
+    instance_port     = 3000
+    instance_protocol = "http"
+    lb_port           = 80
+    lb_protocol       = "http"
+  }
+
+  # Ensure we allow incoming traffic to the ELB, HTTP/S
+  security_groups = ["${module.elb-sg.id}"]
+
+  # ELBs in the public subnets, separate from the web ASG in private subnets
+  subnets = ["${module.vpc.public_subnet_ids}"]
+}
+
+module "web" {
+  source        = "../../modules/asg"
+  ami           = "${module.ubuntu-xenial-ami.id}"
+  azs           = "${local.azs}"
+  name_prefix   = "${var.name}-web"
+  elb_names     = ["${aws_elb.web.name}"]
+  instance_type = "t2.nano"
+  max_nodes     = "${length(module.vpc.public_subnet_ids)}"
+  min_nodes     = "${length(module.vpc.public_subnet_ids)}"
+  public_ip     = false
+  key_name      = "${aws_key_pair.main.key_name}"
+  subnet_ids    = ["${module.vpc.private_subnet_ids}"]
```

Comment:
how are the public and private subnets communicating with each other? have you tested this with the ASG and ELB? From the network description/requirements you shared previously, I was under the impression the public and private subnets could not communicate. The specific details of the topology need to be made more clear in the README. As a reviewer, I should not be wondering if it is A or B and how does C happen, I should be able to find the code that confirms that A, B and C are in fact happening, and are implemented as stated. This is the purpose of the requirements spec being explicit.

Reply:
the ELB is where private and public subnets merged. I base this code in the scenario-2 and it should work because kube-stack-private uses this scenario.